### PR TITLE
[new release] zar (0.9)

### PR DESCRIPTION
**Pull request by bagnalla**
Formally verified sampling from discrete probability distributions

- Project page: <a href="https://github.com/bagnalla/ocamlzar">https://github.com/bagnalla/ocamlzar</a>

##### CHANGES:

* Initial release. Includes coin, die, and findist samplers with
  QCheck tests for numeric conversion functions.

### CODE REVIEW

File: packages/zar/zar.0.9/opam [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+maintainer: ["Alexander Bagnall <abagnalla@gmail.com>"]
+authors: ["Alexander Bagnall"]
+license: "MIT"
+tags: ["zar" "random" "sampler" "verified" "coq" "discrete"]
+homepage: "https://github.com/bagnalla/ocamlzar"
+bug-reports: "https://github.com/bagnalla/ocamlzar/issues"
+depends: [
+  "ocaml"
+  "dune" {>= "3.7"}
+  "odoc" {with-doc}
```
Comment (mseri) [inline packages/zar/zar.0.9/opam:15]:
```suggestion
  "qcheck-alcotest" {with-test"
  "odoc" {with-doc}
```

File: packages/zar/zar.0.9/opam [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+description:
+  "See the [paper](https://arxiv.org/abs/2211.06747) (to appear in PLDI'23) and [Github repository](https://github.com/bagnalla/zar)."
+maintainer: ["Alexander Bagnall <abagnalla@gmail.com>"]
+authors: ["Alexander Bagnall"]
+license: "MIT"
+tags: ["zar" "random" "sampler" "verified" "coq" "discrete"]
+homepage: "https://github.com/bagnalla/ocamlzar"
+bug-reports: "https://github.com/bagnalla/ocamlzar/issues"
+depends: [
+  "ocaml"
```
Comment (mseri) [inline packages/zar/zar.0.9/opam:13]:
```suggestion
  "ocaml" {>= "4.06"}
```
due to
```

#=== ERROR while compiling zar.0.9 ============================================#
# context              2.2.0~alpha~dev | linux/x86_64 | ocaml-base-compiler.4.02.3 | pinned(https://github.com/bagnalla/ocamlzar/releases/download/0.9/zar-0.9.tbz)
# path                 ~/.opam/4.02/.opam-switch/build/zar.0.9
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p zar -j 127 @install
# exit-code            1
# env-file             ~/.opam/log/zar-7-238ce3.env
# output-file          ~/.opam/log/zar-7-238ce3.out
### output ###
# (cd _build/default && /home/opam/.opam/4.02/bin/ocamlc.opt -w -40 -g -bin-annot -I lib/.zar.objs/byte -intf-suffix .ml -no-alias-deps -open Zar -o lib/.zar.objs/byte/zar__Coin.cmo -c -impl lib/coin.ml)
# File "lib/coin.ml", line 18, characters 2-11:
# Error: Unbound value List.init
# (cd _build/default && /home/opam/.opam/4.02/bin/ocamlc.opt -w -40 -g -bin-annot -I lib/.zar.objs/byte -intf-suffix .ml -no-alias-deps -open Zar -o lib/.zar.objs/byte/zar__Die.cmo -c -impl lib/die.ml)
# File "lib/die.ml", line 16, characters 2-11:
# Error: Unbound value List.init
# (cd _build/default && /home/opam/.opam/4.02/bin/ocamlc.opt -w -40 -g -bin-annot -I lib/.zar.objs/byte -intf-suffix .ml -no-alias-deps -open Zar -o lib/.zar.objs/byte/zar__Findist.cmo -c -impl lib/findist.ml)
# File "lib/findist.ml", line 19, characters 2-11:
# Error: Unbound value List.init
# (cd _build/default && /home/opam/.opam/4.02/bin/ocamlopt.opt -w -40 -g -I lib/.zar.objs/byte -I lib/.zar.objs/native -intf-suffix .ml -no-alias-deps -open Zar -o lib/.zar.objs/native/zar__Die.cmx -c -impl lib/die.ml)
# File "lib/die.ml", line 16, characters 2-11:
# Error: Unbound value List.init
# (cd _build/default && /home/opam/.opam/4.02/bin/ocamlopt.opt -w -40 -g -I lib/.zar.objs/byte -I lib/.zar.objs/native -intf-suffix .ml -no-alias-deps -open Zar -o lib/.zar.objs/native/zar__Findist.cmx -c -impl lib/findist.ml)
# File "lib/findist.ml", line 19, characters 2-11:
# Error: Unbound value List.init
# (cd _build/default && /home/opam/.opam/4.02/bin/ocamlopt.opt -w -40 -g -I lib/.zar.objs/byte -I lib/.zar.objs/native -intf-suffix .ml -no-alias-deps -open Zar -o lib/.zar.objs/native/zar__Coin.cmx -c -impl lib/coin.ml)
# File "lib/coin.ml", line 18, characters 2-11:
# Error: Unbound value List.init
```

**Comment by mseri** [comment]
Is this test failure on 32bit systems expected?
```

#=== ERROR while compiling zar.0.9 ============================================#
# context              2.2.0~alpha~dev | linux/arm32 | ocaml-base-compiler.4.14.1 | pinned(https://github.com/bagnalla/ocamlzar/releases/download/0.9/zar-0.9.tbz)
# path                 ~/.opam/4.14/.opam-switch/build/zar.0.9
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p zar -j 79 @install @runtest
# exit-code            1
# env-file             ~/.opam/log/zar-7-713eed.env
# output-file          ~/.opam/log/zar-7-713eed.out
### output ###
# File "test/dune", line 2, characters 7-10:
# 2 |  (name zar)
#            ^^^
# (cd _build/default/test && ./zar.exe)
# qcheck random seed: 333318369
# Testing `zar'.
# This run has ID `FM0LLK7X'.
# 
#   [OK]          zar          0   int_of_z.
#   [OK]          zar          1   z_of_int.
# > [FAIL]        zar          2   z_of_int_int_of_z.
#   [OK]          zar          3   int_of_z_z_of_int.
#   [OK]          zar          4   z_of_int_plus.
#   [OK]          zar          5   int_of_z_plus.
#   [OK]          zar          6   int_of_z_double.
#   [OK]          zar          7   double_z_of_int.
#   [FAIL]        zar          8   lt_int_of_z.
#   [OK]          zar          9   z_of_int_lt.
# 
# ┌──────────────────────────────────────────────────────────────────────────────┐
# │ [FAIL]        zar          2   z_of_int_int_of_z.                            │
# └──────────────────────────────────────────────────────────────────────────────┘
# 
# test `z_of_int_int_of_z` failed on ≥ 1 cases:
# Zneg XI XO XO XI XI XO XI XO XO XO XO XO XI XO XI XI XO XI XO XI XI XO XO XI XI XO XO XO XO XO XH (after 18 shrink steps)
# 
# [exception] test `z_of_int_int_of_z` failed on ≥ 1 cases:
# Zneg XI XO XO XI XI XO XI XO XO XO XO XO XI XO XI XI XO XI XO XI XI XO XO XI XI XO XO XO XO XO XH (after 18 shrink steps)
# 
#             Raised at QCheck2.Test.check_result in file "src/core/QCheck2.ml", line 2007, characters 6-38
#             Called from Alcotest_engine__Core.Make.protect_test.(fun) in file "src/alcotest-engine/core.ml", line 181, characters 17-23
#             Called from Alcotest_engine__Monad.Identity.catch in file "src/alcotest-engine/monad.ml", line 24, characters 31-35
#             
# Logs saved to `~/.opam/4.14/.opam-switch/build/zar.0.9/_build/default/test/_build/_tests/zar/zar.002.output'.
#  ──────────────────────────────────────────────────────────────────────────────
# 
# Full test results in `~/.opam/4.14/.opam-switch/build/zar.0.9/_build/default/test/_build/_tests/zar'.
# 2 failures! in 0.127s. 10 tests run.
```

**Comment by bagnalla** [comment]
No, it's not expected. I'll investigate on my end. I'm also making some improvements to the programming interface and reorganization of the modules. Should I close this PR and submit a new one with the fixes/changes?

**Comment by mseri** [comment]
As you prefer. If it takes a while I suggest to close it, otherwise you can force push over this

**Pull request closed without merging**